### PR TITLE
Create the study_group app

### LINKED ISSUE #41: Missing Study Group app
an application for Study Group functionallity needs to be created, which includes:
* StudyGroup, GroupMember entities need to be created 
*  Data model methods for said entities
* StudyGroupManager  for more complex queries
* Migration files and Model and app registration

### PR DESCRIPTION
**Pull request by idan1356**
Creates the study_group app along with its models, StudyGroup, GroupManager and GroupMember, and configure the project for said additions as follows:

**study_group folder:**
- Start app `study_group`
- Create models `StudyGroup` and `GroupMember` in `study_group/models.py`
- Create model `GroupManager` for class level queries of `StudyGroup`
- Register models to `admin.py`
- Create initial migration file `0001_initial.py`
- Create test data migration `0002_test_data.py` to populate app 

**HighTeach Folder:**
- Install app to `settings.py`

**Tests Folder**
- Create `conftest.py` file for fixtures of `User`, `StudyGroup`, `GroupMember` entities
- Create Tests for StudyGroup app models using said fixtures

Continues:
#42 

Resolves #41

### CODE REVIEW

**Review by shirazyomtov — APPROVED** [review]
LGTM

**Review by shacharlevy — APPROVED** [review]
LGTM

File: study_group/models.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        return self.group_members.count() >= self.capacity
+
+    def get_all_group_members(self):
+        return self.group_members.all()
+
+    # create a study group and add creator to group
+    @classmethod
+    def create(cls, **kwargs):
+        group_owner = kwargs.pop('group_owner', None)
+        if group_owner is None:
```
Comment (natifridman) [inline study_group/models.py:41]: The default is [`blank=False`](https://docs.djangoproject.com/en/4.2/ref/models/fields/#blank), do you still need this check?

**Review by natifridman — CHANGES_REQUESTED** [review]
Please resolve conflicts. Also, you have some methods here that need to be tested, is this going to be in this PR?

File: study_group/models.py (hunk 1 above)
Comment (idan1356) [reply study_group/models.py]: No need for this check

File: tests/study_group/test_group_manager.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+class TestStudyGroupManager:
+    def test_get_all_non_full_groups(self, persist_group, persist_second_group):
+        for i in range(persist_group.capacity):
+            persist_group.join_group(User.objects.create(username=f'user{i}', password='pass'))
+
+        result = StudyGroup.objects.get_all_non_full_groups()
+        assert persist_group not in result
+        assert persist_second_group in result
+
+    def test_search_group_by_keyword(self, persist_group, persist_second_group):
```
Comment (natifridman) [inline tests/study_group/test_group_manager.py:16]: Maybe we can use [parametrize](https://docs.pytest.org/en/6.2.x/parametrize.html) for doing this test?

**Review by natifridman — COMMENTED** [review]
Nice work. Maybe we can add some more tests here:
- What about a test for a user trying to join a group he is already part of? 
- Group owner trying to join his own group?
- `field` over the char limit?
- Negative capacity

**Comment by natifridman** [comment]
We can't see that the tests are passing since there is no Github action for Pytest

File: tests/study_group/test_group_manager.py (hunk 2 above)
Comment (idan1356) [reply tests/study_group/test_group_manager.py:16]: I agree, it would make more sense, I'll change it

**Comment by idan1356** [comment]
Regarding the `field` over char limit and negative capacity, would a test be still necessary if it gets enforced upon save by
the max characters parameter of `CharField` and the `PositiveIntegerField`?

File: tests/study_group/test_group_manager.py (hunk 2 above)
Comment (idan1356) [reply tests/study_group/test_group_manager.py:16]: I couldn't figure a way to do something along the lines of:
`@pytest.mark.parametrize("keyword", "expected", [("this keyword shouldnt exist", []),
                                                     ("math", [persist_group]),
                                                     ("guitar", [persist_second_group]),
                                                     ("profile", [persist_group, persist_second_group])]
                             )`

**Comment by idan1356** [comment]
Regarding the owner joining own group test, there is no prevention from being a group member as an owner.

**Review by Davidco94 — APPROVED** [review]
LGTM!

**Review by shirazyomtov — APPROVED** [review]
LGTM

**Comment by idan1356** [comment]
Made a small change to the test data migration file.
The transaction is now atomic.

**Review by shirazyomtov — APPROVED** [review]
LGTM

**Review by shacharlevy — DISMISSED** [review]
LGTM!

File: study_group/migrations/0002_test_data.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    groups = [
+        StudyGroup(group_owner=owner1, field='Math',
+                   group_description='This is a group for math enthusiasts', capacity=5),
+        StudyGroup(group_owner=owner2, field='Science',
+                   group_description='This is a group for science lovers', capacity=10),
+        StudyGroup(group_owner=owner1, field='History',
+                   group_description='This is a group for history buffs', capacity=7),
+    ]
+
+    for group in groups:
```
[the remark this answers is not in the document]
Comment (idan1356) [reply study_group/migrations/0002_test_data.py]: Done

File: tests/study_group/test_group_manager.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        empty = StudyGroup.objects.search_group_by_keyword("this keyword should not exist")
+        assert not empty
+
+        math_keyword_set = StudyGroup.objects.search_group_by_keyword("cs")
+        print("###################")
+        print(list(math_keyword_set))
+        print([persist_group])
```
Comment (natifridman) [inline tests/study_group/test_group_manager.py:23]: Why do you have prints here?

File: tests/study_group/test_study_group.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+        with pytest.raises(ValueError):
+            persist_group.join_group(persist_user)
+
+    def test_get_all_group_members(self, persist_group, persist_user, persist_second_user):
+        assert not list(persist_group.get_all_group_members())
+
+        persist_group.join_group(persist_user)
+        persist_group.join_group(persist_second_user)
+        print(set(persist_group.get_all_group_members()))
```
Comment (natifridman) [inline tests/study_group/test_study_group.py:35]: same

**Review by natifridman — CHANGES_REQUESTED** [review]

File: tests/study_group/test_group_manager.py (hunk 4 above)
Comment (idan1356) [reply tests/study_group/test_group_manager.py:23]: Sorry, forgot to remove them

File: study_group/migrations/0002_test_data.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
+from django.db import migrations, transaction
+
+
+def generate_data(apps, schema_editor):
+    from django.contrib.auth.models import User
+    from study_group.models import StudyGroup, GroupMember
+
+    owner1 = User.objects.create_user(username='owner1', password='password')
```
Comment (natifridman) [inline study_group/migrations/0002_test_data.py:8]: Make sure not to use these usernames in other migrations

**Review by natifridman — APPROVED** [review]
Overall looks good, added one comment to consider.
Please squash your commits

**Comment by idan1356** [comment]
Why are the two commits need to be squashed? I thought its better to keep each commit with 1 purpose only (logic creation as opposed to logic testing)

**Comment by natifridman** [comment]
> Why are the two commits need to be squashed? I thought its better to keep each commit with 1 purpose only (logic creation as opposed to logic testing)

Sounds good, so let's keep it with the 2 commits.
Waiting for maintainers approval

**Pull request merged**